provisioner/provision.js

```diff
@@ -0,0 +1,302 @@
+var assert = require('assert');
+
+var MAX_PROVISION_ITERATION = 1000 * 60 * 20; // 20 minutes
+
+// Docs for Ec2: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/EC2.html
+
+
+/**
+ * A provisioner object represents our knowledge of how to take AWS state, pricing data
+ * and WorkerType definitions and provision EC2 instances.  It does not understand
+ * how to actually create instances or fetch the state or pricing, rather it defers
+ * to the workerType to create instances and the state and pricing determine all of
+ * the relevant facts needed
+ */
+function Provisioner(cfg) {
+  // This is the ID of the provisioner.  It is used to interogate the queue
+  // for pending tasks
+  assert(cfg.provisionerId);
+  assert(typeof cfg.provisionerId === 'string');
+  this.provisionerId = cfg.provisionerId;
+
+  // This is a prefix which we use in AWS to determine ownership
+  // of a given instance.  If we could tag instances while they were
+  // still spot requests, we wouldn't need to do this.
+  assert(cfg.awsKeyPrefix);
+  assert(typeof cfg.awsKeyPrefix === 'string');
+  this.awsKeyPrefix = cfg.awsKeyPrefix;
+
+  // This is the number of milliseconds to wait between completed provisioning runs
+  assert(cfg.provisionIterationInterval);
+  assert(typeof cfg.provisionIterationInterval === 'number')
+  assert(!isNaN(cfg.provisionIterationInterval));
+  this.provisionIterationInterval = cfg.provisionIterationInterval;
+
+  // This is the Queue object which we use for things like retreiving
+  // the pending jobs.
+  assert(cfg.taskcluster);
+  assert(cfg.taskcluster.credentials)
+
+  // We only grab the credentials for now, no need to store them in this object
+  this.Queue = new taskcluster.Queue({credentials: cfg.taskcluster.credentials});
+
+  // We need a set up WorkerType reference
+  assert(cfg.WorkerType);
+  this.WorkerType = cfg.WorkerType;
+  
+  // We want the subset of AWS regions to use
+  assert(cfg.allowedAwsRegions);
+  this.allowedAwsRegions = cfg.allowedAwsRegions;
+
+  // We need a configured EC2 instance
+  assert(cfg.ec2);
+  this.ec2 = cfg.ec2;
+
+  this.__provRunId = 0;
+
+  // We cache aws pricing data because it's not important to be completely fresh
+  this.pricingCache = new Cache(15, awsPricing, this.ec2);
+
+  // We track outstanding AWS requests here
+  this.__outstanding = {};
+}
+
+module.exports.Provisioner = Provisioner;
+
+/**
+ * Start running a provisioner.
+ */
+Provisioner.prototype.run = function () {
+  var that = this;
+
+  this.__keepRunning = true;
+
+  function provisionIteration() {
+    // We should cancel the last iteration's watch dog
+    if (that.__watchdog) {
+      clearTimeout(that.__watchdog);
+    }
+    // And make sure we set this one!
+    that.__watchdog = setTimeout(function() {
+      generalDebug('KILLING PROVISIONER BECAUSE IT APPEARS TO BE STUCK...');
+      // Hmm, should I instead just process.exit(1);
+      throw new Error('PROVISIONER HAS FALLEN AND CAN\'T GET BACK UP');
```

Comment:
This is pretty neat... I usually include the string `[alert-operator]` in debug messages when I want to be notified...Then I configure an alart on papertail.

Anyways, this is awesome! :)